python-package/lightgbm/sklearn.py

```diff
@@ -483,6 +490,33 @@ def _extract_evaluation_meta_data(
         raise TypeError(f"{name} should be dict or list")
 
 
+def _validate_eval_set_Xy(eval_set, eval_X, eval_y):
+    """Validate eval args.
+
+    Returns
+    -------
+    eval_set
+    """
+    if eval_set is not None:
+        msg = "The argument 'eval_set' is deprecated, use 'eval_X' and 'eval_y' instead."
+        warnings.warn(msg, category=LGBMDeprecationWarning, stacklevel=2)
+        if eval_X is not None or eval_y is not None:
+            raise ValueError("Specify either 'eval_set' or 'eval_X' and 'eval_y', but not both.")
+        return eval_set
+    if (eval_X is None) != (eval_y is None):
+        raise ValueError("You must specify eval_X and eval_y, not just one of them.")
+    if eval_set is None and eval_X is not None:
+        if isinstance(eval_X, tuple) != isinstance(eval_y, tuple):
+            raise ValueError("If eval_X is a tuple, y_val must be a tuple of same length, and vice versa.")
+        if isinstance(eval_X, tuple) and len(eval_X) != len(eval_y):
+            raise ValueError("If eval_X is a tuple, y_val must be a tuple of same length, and vice versa.")
+        if isinstance(eval_X, tuple):
+            eval_set = list(zip(eval_X, eval_y))
+        else:
+            eval_set = [(eval_X, eval_y)]
+    return eval_set
```

Comment:
Can you please add test cases covering all these `if-else` branches? e.g. using `pytest.raises()` for the error cases and checking that the expected error type + message are raised?

There are notes in #7031 on how to generate a coverage report with `pytest-cov` for LightGBM's python tests if you want to use that to double-check.

Reply:
`sklearn.py` has covered everything modified by this PR.

Reply:
new tests look great, thank you!